File: patterns/chainofresponsibility.js
//chain of responsibility pattern
//pass request along chain of objects

// Define the Handler interface, which specifies the method for handling requests
class Handler {
    constructor() {
        this.nextHandler = null;
    }

    setNext(handler) {
        this.nextHandler = handler;
    }

    handle(request) {
        if (this.nextHandler) {
            return this.nextHandler.handle(request);
        }

        return null;
    }
}

// Define the Concrete Handlers, which implement the Handler interface and handle specific requests
class ConcreteHandler1 extends Handler {
    handle(request) {
        if (request === "request1") {
            console.log("ConcreteHandler1: Handling request1");
            return true;
        } else {
            return super.handle(request);
        }
    }
}

class ConcreteHandler2 extends Handler {
    handle(request) {
        if (request === "request2") {
            console.log("ConcreteHandler2: Handling request2");
            return true;
        } else {
            return super.handle(request);
        }
    }
}

class ConcreteHandler3 extends Handler {
    handle(request) {
        if (request === "request3") {
            console.log("ConcreteHandler3: Handling request3");
            return true;
        } else {
            return super.handle(request);
        }
    }
}

// Create the chain of responsibility by linking the handlers together
const handler1 = new ConcreteHandler1();
const handler2 = new ConcreteHandler2();
const handler3 = new ConcreteHandler3();

handler1.setNext(handler2);
handler2.setNext(handler3);

// Send requests through the chain of responsibility
handler1.handle("request1");
handler1.handle("request2");
handler1.handle("request3");

File: patterns/chainofresponsibility.py
# Define the Handler interface, which defines the interface for handling requests
class Handler:
    def set_next(self, handler):
        pass

    def handle(self, request):
        pass

# Define the AbstractHandler class, which implements the Handler interface and provides a default implementation of set_next
class AbstractHandler(Handler):
    def __init__(self):
        self._next_handler = None

    def set_next(self, handler):
        self._next_handler = handler
        return handler

    def handle(self, request):
        if self._next_handler is not None:
            return self._next_handler.handle(request)
        else:
            return None

# Define the ConcreteHandler class, which implements the Handler interface and handles requests
class ConcreteHandler1(AbstractHandler):
    def handle(self, request):
        if request == "handler1":
            return f"Handled by {type(self).__name__}"
        else:
            return super().handle(request)

class ConcreteHandler2(AbstractHandler):
    def handle(self, request):
        if request == "handler2":
            return f"Handled by {type(self).__name__}"
        else:
            return super().handle(request)

class ConcreteHandler3(AbstractHandler):
    def handle(self, request):
        if request == "handler3":
            return f"Handled by {type(self).__name__}"
        else:
            return super().handle(request)

# Use the Handler classes to handle requests
handler1 = ConcreteHandler1()
handler2 = ConcreteHandler2()
handler3 = ConcreteHandler3()

handler1.set_next(handler2).set_next(handler3)

print(handler1.handle("handler1"))
print(handler1.handle("handler2"))
print(handler1.handle("handler3"))
